File: kernel/tools/patchin.pl
#!/usr/bin/perl -w
#About this script:
#
#It was written to avoid having to modify the unreadable
#monster that is patchin.sh. Shell scripts should not be
#that long.
#
#Anyway, this is heavily commented and uses perl, which
#is almost as prevalent as bash on Linux systems and
#seeing as how the kernel itself is dependent on perl for
#a number of its utilities, it's probably a safe bet to
#be able to rely on it being present. Requires perl 5.10.
use Switch;
use File::Copy;
use Fcntl ':mode';
use Tie::File;
use FindBin qw($Bin);
use Cwd;

$cwd = getcwd;
my $Base = join("/",$Bin,"..");
my $majver = 1;
my $minver = 1;
my $subver = 1;
my $kernelver = "";

sub print_help
{
	print "Patchin Script\n";
	print "Run the patchin script while in the root of the kernel\n";
	print "you wish to patch into. Paths will be determined automatically.\n";
	print "By default, patchin will use softlinks for all files.\n";
	print "\n";
	print "Usage: patchin.pl [OPTION] [PATH]\n";
	print " --copy			Add via copy.\n";
	print " --link			Add via softlink.\n";
	print " --stock			Some older kernel need patches from stock just to build.\n";
	print " --assume-yes		Assume yes to all questions\n";
	print " --assume-subver <NUM> 	Assume we are patching against kernel minor version NUM\n";
	print "\n";
	print "[PATH] should point to a kernel directory, otherwise the current\n";
	print "working directory will be assumed to be the desired kernel directory\n";
}

sub parse_args
{
	foreach $arg (0 .. $#ARGV) {
		switch ($ARGV[$arg]) {
			my $opt = $ARGV[$arg];
			case ("--copy") { $insert_type = "copy"; }
			case ("--link") { $insert_type = "link"; }
			case ("--help") { print_help(); exit;}
			case ("--assume-yes") { $assume_yes = 1; }
			case ("--assume-subver") { $assume_subver = 1; }
			case ("--stock") { $stock = 1; }
			case (/^\d*$/) { if ($assume_subver) { $assume_subver = $opt; } }
			case (/^[\/|\.|\w].*/) { $path = $opt; }
			else { print "don't know option $opt\n"; }
		}
	}
}

sub get_kernel_version
{
	my $file = open(FH, join("/",$path,"Makefile"));
	if ($file) {
		my $maj_vstr = readline(FH);
		my $min_vstr = readline(FH);
		my $sub_vstr = readline(FH);
		close(FH);
		chomp($majver = (split(/ /, $maj_vstr))[2]);
		chomp($minver = (split(/ /, $min_vstr))[2]);
		chomp($subver = (split(/ /, $sub_vstr))[2]);
		$kernelver = "v".$majver.".".$minver.".".$subver
	}
}

sub set_target
{
	if ($path eq "") {
		$path = $cwd;
	}
}

# 2 parameters
# - Identifier to look for
# - Array to find it in
#
# We handle the parameters oddly to force
# the second to be recognized as an array.
sub do_find_index_in_file
{
	my $ind = -1;
	my ($id, @lines) = @_;
	foreach (0 .. $#lines) {
		if ($lines[$_] =~ m/(\Q$id\E)/ ) {
			$ind = $_;
		}
	}
	return $ind;
}

# 3 parameters
# - File to source, and file to patch
# - Identifier for line after which the line is to be
#   inserted/
# - Line to patch into the file.
sub do_patch_line
{
	my $expath = join("/",$path,$_[0]);
	print "Patching file $expath\n";
	# Tie is pretty nifty.
	tie my @lines, 'Tie::File', $expath;
	my $id = $_[1];
	my $insert = $_[2];
	my $index = -1;

	$index = do_find_index_in_file($insert,@lines);
	if ($index > -1) {
		die "File $expath already patched.\n";
	}
	
	$index = do_find_index_in_file($id,@lines);
	if ($index == -1) {
		die "Couldn't find $id in $expath\n";
	}

	splice(@lines,($index + 1), 0, "$insert");
	untie @lines;
}

# 2 parameters
# - File to source, and file to patch
# - Identifier for line after which the contents of the
#   source file are to be placed.
sub do_patch_file
{
	my $id = $_[1];
	my $expath = join("/",$path,$_[0]);
	my $srcpath = join("/",$Base,$_[0]);

	print "Splicing file $srcpath into $expath\n";

	tie my @existing, 'Tie::File', $expath;
	tie my @source, 'Tie::File', $srcpath;

	if (do_find_index_in_file($source[0],@existing) > -1) {
		die "File $expath is already spliced\n";
	}

	$index = do_find_index_in_file($id,@existing);

	splice(@existing,($index),0,@source);

	untie @existing;
	untie @source;
}

sub link_file
{
	print "Linking file: $_[0] -> $_[1]\n";
	symlink($_[0],$_[1]);
}

sub copy_file
{
	print "Copying file: $_[0] -> $_[1]\n";
	copy($_[0],$_[1]);
}

sub do_process_dir
{
	my $filedir = $_[0];
	my $fileroot = join("/",$Base,$filedir);
	my $filedest = join("/",$path,$filedir);

	if ( ! -e $filedest) {
		mkdir($filedest);
	}

	opendir(INDIR, $fileroot);
	while ( defined ( my $file = readdir(INDIR))) {
		if (grep(/^\.\.*$/,$file) > 0) {
			next;
		}
		$src = join("/",$fileroot,$file);
		$dst = join("/",$filedest,$file);
		if ( -d $src ) {
			do_process_dir(join("/",$filedir,$file));
			next;
		}
		switch($insert_type) {
			case ("link") { link_file($src,$dst); }
			case ("copy") { copy_file($src,$dst); }
		}
	}
}

# When given a directory, will ensure that the directory named
# exists in the target location. It will then either link or
# copy every file in that directory to the other directory.
sub do_insert_files
{
	my $filedir = $_[0];
	do_process_dir($filedir);
}

# Looks for patches that associate with this kernel version
# and then applies them to our kernel.
sub do_kernel_patches
{
	my $expath = $path;
	my $srcpath = join("/",$Base,$_[0],$kernelver);
	my @files;

	opendir(DIR, $srcpath);
	while (my $file = readdir(DIR)) {
		# We only want files
		next unless (-f "$srcpath/$file");
		push @files, "$srcpath/$file";
	}
	closedir(DIR);

	foreach (sort @files) {
		system("patch -p1 -i $_ -d $path") == 0
		or die "system @args failed: $?"
	}
}

$insert_type = "link";
$path = "";
$patching_project = "AXFS";
$assume_yes = 0;
$stock = 0;

parse_args();
set_target();
get_kernel_version();

if ($stock == 1) {
	do_kernel_patches("patches/stock_patches");
	exit;
}

if ($assume_yes != 1) {
	print "Patching $patching_project into linux-$majver.$minver.$subver.\n";
	print "Using method: $insert_type\n";
	print "Proceed? [Y/n]: ";
	$resp = readline(*STDIN);

	if (grep(/[n|N]/,$resp) > 0) {
		exit;
	}
}

if ($minver > 5 || $assume_subver > 5) {
	do_kernel_patches("patches");
	do_patch_line("fs/Makefile", "CONFIG_CRAMFS", "obj-\$(CONFIG_AXFS)		+= axfs/");
	do_patch_file("fs/Kconfig", "config ECRYPT_FS");
	do_insert_files("fs/axfs");
	do_insert_files("include/linux");
}

File: kernel/tools/patchin.rb
#!/usr/bin/ruby
require 'fileutils'
require "pp"

def get_kernel_version(config)
  file = File.join("/",config[:path],"Makefile")
  if not File.file?(file) then raise "#{config[:path]} not a valid kernel directory." end
  lines = File.readlines(file)
  config[:majver] = lines[0].split(" ")[2]
  config[:minver] = lines[1].split(" ")[2]
  config[:subver] = lines[2].split(" ")[2]
  if (config[:subver] == "0")
    config[:kernelver] = "v" + config[:majver] + "." + config[:minver]
  else
    config[:kernelver] = "v" + config[:majver] + "." + config[:minver] + "." + config[:subver]
  end
end

def print_help
  puts "Patchin Script"
  puts ""
  puts "Usage: patchin.rb [OPTION] [PATH]"
  puts " --copy			Add via copy. [default: use symlinks]"
  puts " --none			Skip AXFS files, patches only."
  puts " --stock			Apply patches we've added to fix older kernels build with newer tools. Then quit."
  puts " --assume-yes		Assume yes to all questions [default: ask user]"
  puts " --pretend		Don't apply any changes, just pretend you're going to."
  puts ""
  puts "[PATH] should point to a kernel directory, otherwise the current"
  puts "working directory will be assumed to be the desired kernel directory"
end

def parse_args()
  config = {}
  config[:cwd] = Dir.getwd
  config[:base] = File.expand_path(File.dirname(__FILE__))
  config[:majver ]= 1
  config[:minver] = 1
  config[:subver] = 1
  config[:kernelver] = ""

  config[:insert_type] = "link"
  config[:path] = config[:cwd]
  config[:assume_yes] = false
  config[:stock] = false
  config[:pretend] = false


  ARGV.each do |arg|
    if (arg == "--copy")
      config[:insert_type] = "copy"
    elsif (arg == "--none")
      config[:insert_type] = "none"
    elsif (arg == "--help")
      print_help()
      exit
    elsif (arg == "--assume-yes")
      config[:assume_yes] = true
    elsif (arg == "--stock")
      config[:stock] = true
    elsif (arg == "--pretend")
      config[:pretend] = true
    elsif (arg =~ /^[~|\/|\.|\w].*/)
      config[:path] = arg
    else
      raise "don't know option #{arg}"
    end
  end

  get_kernel_version(config)

  config
end

def select_files(path)
  files = []
  Dir.entries(path).each do |f|
    file = File.join(path,f)
    if not File.file?(file) then next end
    files.push(f)
  end
  files
end

def patch_kernel(config,in_path)
  base_path = File.expand_path(File.join(config[:base],"..",in_path))
  if not File.directory?(base_path) then raise "Invalid directory - #{base_path}" end
  path = File.join(base_path,config[:kernelver])
  if not File.directory?(path)
    puts "nothing to patch for #{config[:kernelver]}"
    return
  end
  select_files(path).each do |patch|
    cmd = "patch -p1 -i #{File.join(path,patch)} -d #{config[:path]}"
    if config[:pretend] then cmd += " --dry-run" end
    puts "Applying patch - #{patch}"
    #puts "cmd: \"#{cmd}\""
    if not system(cmd)
      raise "cmd: \"#{cmd}\" failed with #{$?}"
    end
  end
end

def insert_files(config,in_path)
  srcdir = File.expand_path(File.join(config[:base],"..",in_path))
  if not File.directory?(srcdir) then raise "Invalid directory - #{srcdir}" end
  dstdir = File.expand_path(File.join(config[:path],in_path))
  if not config[:pretend]
    FileUtils.mkdir_p(dstdir)
  end
  select_files(srcdir).each do |f|
    srcfile = File.join(srcdir,f)
    dstfile = File.join(dstdir,f)
    puts "Inserting #{f} via #{config[:insert_type]}"
    if config[:pretend] then next end
    if config[:insert_type] == "link"
      FileUtils.ln_s(srcfile,dstfile)
    else
      FileUtils.cp(srcfile,dstfile)
    end
  end
end

def insert_line(config,path,tokens,newline)
  file = File.join(config[:path],path)
  if not File.file?(file) then raise "#{file} missing - not a valid kernel directory." end
  filedata = File.read(file)
  token = ""
  tokens.each do |t|
    if filedata =~ /#{t}/
      token = t
      break
    end
  end
  if token == "" then raise "couldn't find tokens #{tokens.join(",")} in #{file}" end
  newdata = []
  filedata.split("\n").each do |line|
    newdata.push(line)
    if line =~ /#{token}/
      puts "Inserting \"#{newline}\" after \"#{line}\""
      newdata.push(newline)
    end
  end
  if config[:pretend] then return end
  File.open(file, 'w') {|f| f.write(newdata.join("\n")+"\n") }
end

config = parse_args()

pp config


if config[:stock]
  #patches we apply just to get the stock kernel to build with modern build tools
  puts "Patching linux-#{config[:majver]}.#{config[:minver]}.#{config[:subver]} for building.";
  patch_kernel(config,"patches/stock_patches")
  exit
end

if not config[:assume_yes]
  puts "Patching AXFS into linux-#{config[:majver]}.#{config[:minver]}.#{config[:subver]}.";
  puts "Using method: #{config[:insert_type]}";
  print "Proceed? [Y/n]: ";
  resp = STDIN.gets

  if resp =~ /[n|N]/
    puts "Goodbye!"
    exit
  end
end

#patches to kernel outside of axfs code
patch_kernel(config,"patches")

if config[:insert_type] == "none" then exit end
insert_line(config,"fs/Makefile", ["CONFIG_SQUASHFS","CONFIG_CRAMFS"], "obj-\$(CONFIG_AXFS)		+= axfs/");
insert_line(config,"fs/Kconfig",["fs/squashfs","fs/cramfs","config ECRYPT_FS"],"source \"fs/axfs/Kconfig\"");
insert_files(config,"fs/axfs");
insert_files(config,"include/linux");
